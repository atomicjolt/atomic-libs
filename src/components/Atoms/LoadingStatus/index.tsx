import React from "react";
import ThreeDotLoader from "../../Loaders/ThreeDotLoader";
import { ErrorBanner } from "../../Banners/DismissableBanner";
import { LoadingMessage, LoadingWrapper } from "./LoadingStatus.styles";

export interface LoadingStatusProps {
  /** Loading status, when true, a loading animation is displayed  */
  readonly isLoading?: boolean;
  /** Optional message to display beneath the loading animation */
  readonly loadingMessage?: React.ReactNode;
  /** Customize what is rendered when in a loading state */
  readonly renderLoading?: (
    isLoading: boolean,
    message: React.ReactNode
  ) => React.ReactNode;
  /** An error. When present, an error banner will be displayed */
  readonly error?: React.ReactNode;
  /** Cutomsize what is rendered when in an error state */
  readonly renderError?: (error: React.ReactNode) => React.ReactNode;
  /** If `loading` is false and `error` is absent, the children will be rendered */
  readonly children?: React.ReactNode;
}

const renderLoadingDefault = (isLoading: boolean, message: React.ReactNode) => {
  return (
    <LoadingWrapper className="aje-loading-status-loading">
      <div>
        <ThreeDotLoader isLoading={isLoading} />
        {message && <LoadingMessage>{message}</LoadingMessage>}
      </div>
    </LoadingWrapper>
  );
};

const renderErrorDefault = (error: React.ReactNode) => {
  return <ErrorBanner>{error}</ErrorBanner>;
};

/** Combination of a loading and error component.
 * Useful when waiting for a resource to load.
 *
 * **Note**: If both an `error` is present and `loading` is true,
 * the error takes precedencs and will be displayed  */
export default function LoadingStatus(props: LoadingStatusProps) {
  const {
    isLoading = false,
    loadingMessage = null,
    renderLoading = renderLoadingDefault,
    error = null,
    renderError = renderErrorDefault,
    children = null,
  } = props;

  if (error) {
    return <>{renderError(error)}</>;
  }

  if (isLoading) {
    return <>{renderLoading(isLoading, loadingMessage)}</>;
  }

  return <>{children}</>;
}
