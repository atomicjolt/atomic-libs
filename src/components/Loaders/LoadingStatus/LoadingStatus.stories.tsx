import { Meta, StoryObj } from "@storybook/react";
import { getCssProps } from "@sb/cssprops";
import { SpinnerLoader } from "../SpinnerLoader";
import { Banner } from "@components/Banners/Banner";
import { LoadingStatus } from ".";

const meta: Meta<typeof LoadingStatus> = {
  title: "Loaders/LoadingStatus",
  component: LoadingStatus,
  tags: ["!autodocs"],
  parameters: {
    cssprops: getCssProps("Loader"),
  },
  argTypes: {
    loadingMessage: {
      type: "string",
    },
    error: {
      type: "string",
    },
  },
};

export default meta;

type Story = StoryObj<typeof LoadingStatus>;

export const Primary: Story = {
  args: {
    isLoading: true,
    loadingMessage: "Loading...",
    error: "",
    children: "This is the content",
  },
};

export const Error: Story = {
  args: {
    isLoading: false,
    loadingMessage: "",
    error: "An error occurred",
    children: "This is the content",
  },
};

export const CustomizeLoadingState: Story = {
  args: {
    isLoading: true,
    loadingMessage: "Loading...",
    error: "",
    children: "This is the content",
    renderLoading: <SpinnerLoader />,
  },
};

export const CustomizeErrorState: Story = {
  args: {
    isLoading: false,
    loadingMessage: "",
    error: "An error occurred",
    children: "This is the content",
    renderError: <div>Error!</div>,
  },
};

export const WithData: Story = {
  args: {
    children: (data: any) => <Banner variant="info">{data}</Banner>,
    data: "This is the data!",
  },
  parameters: {
    docs: {
      source: {
        code: `
<LoadingStatus data="This is the data!">
  {(data) => <Banner variant="info">{data}</Banner>}
</LoadingStatus>
`,
      },
    },
  },
};

export const NoDataFallback: Story = {
  args: {
    data: null,
    children: (data: any) => <Banner variant="info">{data}</Banner>,
    fallback: <Banner variant="warning">No data available</Banner>,
  },
  parameters: {
    docs: {
      source: {
        code: `
<LoadingStatus data={null} fallback={<Banner variant="warning">No data available</Banner>}>
  {(data) => <Banner variant="info">{data}</Banner>}
</LoadingStatus>
        `,
      },
    },
  },
};
