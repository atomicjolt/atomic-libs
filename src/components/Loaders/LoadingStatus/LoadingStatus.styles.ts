import styled from "styled-components";
import mixins from "../../../styles/mixins";

export const LoadingWrapper = styled.div`
  --loader-size: 20px;

  position: absolute;
  inset: 0;
  display: grid;
  place-items: center;

  .aje-three-dot-loader {
    position: relative;
  }
`;

export const LoadingContent = styled.div`
  display: flex;
  flex-direction: column;
  place-items: center;
`;

export const LoadingMessage = styled.p`
  ${mixins.Regular}
  font-size: var(--loader-size);
  color: var(--text-clr);
`;
