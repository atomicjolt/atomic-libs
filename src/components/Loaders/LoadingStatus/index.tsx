import React from "react";
import ThreeDotLoader from "../ThreeDotLoader";
import { ErrorBanner } from "../../Banners/DismissableBanner";
import {
  ErrorStateProps,
  LoaderPlacement,
  LoaderProps,
} from "../Loading.types";

export interface LoadingStatusProps extends LoaderProps {
  /** Loading status, when true, a loading animation is displayed  */
  readonly isLoading?: boolean;

  /** Optional message to display beneath the loading animation */
  readonly loadingMessage?: React.ReactNode;

  /** Placement of the loader */
  readonly loadingPlacement?: LoaderPlacement;

  /** Customize what is rendered when in a loading state */
  readonly renderLoading?: React.ReactNode | React.ComponentType<LoaderProps>;

  /** An error. When present, an error banner will be displayed */
  readonly error?: React.ReactNode;
  /** Customize what is rendered when in an error state */
  readonly renderError?: React.ReactNode | React.ComponentType<ErrorStateProps>;

  /** If `loading` is false and `error` is absent, the children will be rendered */
  readonly children?: React.ReactNode;
}

const ErrorDefault = (props: ErrorStateProps) => {
  const { error } = props;
  return <ErrorBanner>{error}</ErrorBanner>;
};

/** Combination of a loading and error component.
 * Useful when waiting for a resource to load.
 *
 * **Note**: If both an `error` is present and `loading` is true,
 * the error takes precedencs and will be displayed  */
export function LoadingStatus(props: LoadingStatusProps) {
  const {
    isLoading = false,
    renderLoading: Loading = ThreeDotLoader,
    loadingMessage = null,
    loadingPlacement = "inline",
    error = null,
    renderError: Error = ErrorDefault,
    children = null,
  } = props;

  if (error) {
    return typeof Error === "function" ? <Error error={error} /> : Error;
  }

  if (isLoading) {
    return typeof Loading === "function" ? (
      <Loading
        isLoading={isLoading}
        message={loadingMessage}
        placement={loadingPlacement}
      />
    ) : (
      Loading
    );
  }

  return children;
}

export default LoadingStatus;
