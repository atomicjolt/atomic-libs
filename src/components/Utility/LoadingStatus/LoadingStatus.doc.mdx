import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  Stories,
} from "@storybook/addon-docs";

<Title />
<Subtitle />
<Description />
<Primary />
<ArgsTable />
<Stories />

# Example

`LoadingStatus` works very well with things like Apollo

```tsx
import React from "react";
import { useQuery } from "@apollo/client";
import { LoadingStatus } from "atomic-elements";

export default function PageLoader() {
  const { loading, error, data } = useQuery(SOME_QUERY);

  return (
    <LoadingStatus loading={loading} error={error.message}>
      <Page data={data} />
    </LoadingStatus>
  );
}
```
