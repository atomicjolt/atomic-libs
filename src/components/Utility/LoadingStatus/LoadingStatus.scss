@use '../../mixins.scss';

.aj-loading-status-loading {
  --loader-size: 20px;

  position: absolute;
  inset: 0;
  display: grid;
  place-items: center;

  .aj-three-dot-loader {
    position: relative;
  }

  p {
    @include mixins.regular;
    font-size: var(--loader-size);
    color: var(--text-clr);
  }
}
