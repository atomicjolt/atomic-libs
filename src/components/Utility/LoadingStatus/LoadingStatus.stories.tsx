import { ComponentMeta, ComponentStory } from "@storybook/react";
import LoadingStatus, { Props } from ".";
import Doc from "./LoadingStatus.doc.mdx";

export default {
  title: "Utility/LoadingStatus",
  component: LoadingStatus,
  parameters: {
    docs: { page: Doc },
  },
  argTypes: {
    error: {
      control: {
        type: "text",
      },
    },
    loadingMessage: {
      control: {
        type: "text",
      },
    },
    children: {
      control: false,
    },
  },
} as ComponentMeta<typeof LoadingStatus>;

const Template: ComponentStory<typeof LoadingStatus> = (args: Props) => (
  <LoadingStatus {...args}>
    <div>Here's some Content that will render when loading completes</div>
  </LoadingStatus>
);

export const Default = Template.bind({});
Default.args = {
  loading: true,
  loadingMessage: "Loading, please wait...",
  error: "",
};

// Additional stories go here
