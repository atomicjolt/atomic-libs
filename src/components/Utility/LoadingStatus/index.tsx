import React from "react";
import ThreeDotLoader from "../../Loaders/ThreeDotLoader";
import { ErrorBanner } from "../../Banners/DismissableBanner";
import { LoadingMessage, LoadingWrapper } from "./LoadingStatus.types";

export interface LoadingStatusProps {
  /** Loading status, when true, a loading animation is displayed  */
  readonly loading?: boolean;
  /** Optinal message to display beneath the loading animation */
  readonly loadingMessage?: React.ReactNode;
  /** Customize what is rendered when in a loading state */
  readonly renderLoading?: (
    loading: boolean,
    message: React.ReactNode
  ) => React.ReactNode;
  /** An error. When present, an error banner will be displayed */
  readonly error?: React.ReactNode;
  /** Cutomsize what is rendered when in an error state */
  readonly renderError?: (error: React.ReactNode) => React.ReactNode;
  /** If `loading` is false and `error` is absent, the children will be rendered */
  readonly children?: React.ReactNode;
}

const renderLoadingDefault = (loading: boolean, message: React.ReactNode) => {
  return (
    <LoadingWrapper className="aje-loading-status-loading">
      <div>
        <ThreeDotLoader loading={loading} />
        {message && <LoadingMessage>{message}</LoadingMessage>}
      </div>
    </LoadingWrapper>
  );
};

const renderErrorDefault = (error: React.ReactNode) => {
  return <ErrorBanner>{error}</ErrorBanner>;
};

/** Combination of a loading and error component.
 * Useful when waiting for a resource to load.
 *
 * **Note**: If both an `error` is present and `loading` is true,
 * the error takes precedencs and will be displayed  */
export default function LoadingStatus(props: LoadingStatusProps) {
  const {
    loading = false,
    loadingMessage = null,
    renderLoading = renderLoadingDefault,
    error = null,
    renderError = renderErrorDefault,
    children = null,
  } = props;

  if (error) {
    return <>{renderError(error)}</>;
  }

  if (loading) {
    return <>{renderLoading(loading, loadingMessage)}</>;
  }

  return <>{children}</>;
}
