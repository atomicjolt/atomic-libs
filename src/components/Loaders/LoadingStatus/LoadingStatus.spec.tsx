import React from "react";
import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import LoadingStatus from ".";

test("matches snapshot", () => {
  const result = render(<LoadingStatus />);
  expect(result.asFragment()).toMatchSnapshot();
});
