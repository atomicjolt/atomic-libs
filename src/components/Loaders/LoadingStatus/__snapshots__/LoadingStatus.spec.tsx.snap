// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadingStatus > matches error snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dhKdPU coQQBZ aje-banner--error"
  >
    <i
      aria-hidden="true"
      class="sc-jXbVAB jVtNFG material-icons is-medium"
    >
      error
    </i>
    <div
      class="sc-kpDprT cnhKCE"
    >
      An error occurred
    </div>
    <button
      aria-label="dismiss error"
      class="sc-cPiJYC sc-dLNtp fTQRDG kjiHHG aje-btn aje-btn--icon aje-banner__dismiss aje-btn--banner"
      role="link"
      tabindex="0"
      type="button"
    >
      <i
        aria-hidden="true"
        class="sc-jXbVAB jVtNFG material-icons is-medium"
      >
        close
      </i>
    </button>
  </div>
</DocumentFragment>
`;

exports[`LoadingStatus > matches loading snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsFSjX kMYkab aje-three-dot-loader is-medium"
    data-placement="center"
  >
    <div
      class="sc-imWZod dRvpRG aje-flex"
    >
      <div
        class="sc-aYaIB bjNylc dots-container"
      >
        <span
          class="sc-gEvDqW sc-eqUzNf iZkYKK gEoPMl dot1"
        />
        <span
          class="sc-gEvDqW sc-fqkwJk iZkYKK hcdaYo dot2"
        />
        <span
          class="sc-gEvDqW sc-dcJtft iZkYKK hqCEWn dot3"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
